massa-models/src/address.rs

```diff
@@ -179,202 +276,606 @@ impl FromStr for Address {
-    pub fn from_prefixed_bytes(data: &[u8]) -> Result<Address, ModelsError> {
-        let deser = AddressDeserializer::new();
-        let (_, res) = deser
-            .deserialize::<DeserializeError>(data)
-            .map_err(|_| ModelsError::AddressParseError)?;
-        Ok(res)
+    fn from_bytes_without_version(data: &[u8]) -> Result<SCAddress, ModelsError> {
+        Ok(SCAddress {
+            hash: Hash::from_bytes(&data.try_into().map_err(|_| {
+                ModelsError::BufferError(format!(
+                    "expected a buffer of size {}, but found a size of {}",
+                    HASH_SIZE_BYTES,
+                    &data.len()
+                ))
+            })?),
+        })
+    }
+}
+
+#[transition::impl_version(versions("1"))]
+impl SCAddress {
+    /// Serialize the address as bytes. Includes the type and version prefixes
+    pub fn to_prefixed_bytes(self) -> Vec<u8> {
+        let mut buff = vec![];
+        let addr_type_ser = U64VarIntSerializer::new();
+        let addr_vers_ser = U64VarIntSerializer::new();
+        addr_type_ser
+            .serialize(&SC_PREFIX, &mut buff)
+            .expect("impl always returns Ok(())");
+        addr_vers_ser
+            .serialize(&Self::VERSION, &mut buff)
+            .expect("impl always returns Ok(())");
+        buff.extend_from_slice(&self.slot.to_bytes_key()[..]);
+        buff.extend_from_slice(&self.created_addr_index.to_be_bytes()[..]);
+        if self.read_only {
+            buff.push(0u8);
+        } else {
+            buff.push(1u8);
+        }
+        buff
+    }
+
+    fn from_bytes_without_version(data: &[u8]) -> Result<SCAddress, ModelsError> {
+        if data.len() != SLOT_KEY_SIZE + 9 {
+            return Err(ModelsError::BufferError(format!(
+                "expected a buffer of size  than {}, but found a size of {}",
+                SLOT_KEY_SIZE + 9,
+                &data.len()
+            )));
+        }
+
+        let err = Err(ModelsError::BufferError(String::from(
+            "could not deserialize buffer from_bytes_without_version",
+        )));
+
+        let (data_slot, rest) = data.split_at(SLOT_KEY_SIZE);
+        let (data_created_addr_index, data_read_only) = rest.split_at(rest.len() - 1);
+
+        let slot = Slot::from_bytes_key(&data_slot.try_into().expect("Unexpected error"));
+        let created_addr_index = u64::from_be_bytes(
+            data_created_addr_index
+                .try_into()
+                .expect("Unexpected error"),
+        );
+        let read_only = match data_read_only {
+            [0_u8] => false,
+            [1_u8] => true,
+            _ => {
+                return err;
+            }
+        };
+
+        Ok(SCAddress {
+            slot,
+            created_addr_index,
+            read_only,
+        })
     }
 }
 
+/* /!\ SCAddressV1 not prehashed! */
+impl PreHashed for Address {}
+
```

Comment:
This is probably dangerous. If the SCV1 being marked as prehashed, but actually isn't, then it could cause really painful bugs in the future.